### DISCUSSION TITLE
Rezact Stores - only readable, only writable?

### DISCUSSION THREAD

**Discussion by look997** [Rezact/Rezact#8] [Q&A]
Have you thought about how to make the store read-only or write-only?

How would this be done in Rezact?

PS Preliminary Rezact I like very much, so I ask for details.

**Comment by zachlankton** [comment]
Excellent question, I was just watching a video on svelte about this and was wondering the best way to implement this.
Although I must admit I'm really not sure what the use case for this is as I've never used or created a readonly store (or write only store) in any framework before.

 How do you feel about something like this?

```tsx
export let $someStore = "someValue"

const internalWriteFunction = $someStore.setToReadOnly();

internalWriteFunction("new value")
```

In the example above the store gets set to readonly and is enforced by the framework.  The `setToReadOnly()` method also returns a function to use for internal module use.

I could likely do something similar for a write only store, but please enlighten me, what is the purpose of a write only store?  Or more generically a store that holds a value that can never be read? Or am I misunderstanding?

Thanks for the question!!

**Reply by look997** [reply]
> In the example above the store gets set to readonly and is enforced by the framework. The setToReadOnly() method also returns a function to use for internal module use.

The readonly in Svelte is that there is an initial function that is the only one that can set the value, such as with setTimeout. This is the kind of store that you never need to set from outside.

> I could likely do something similar for a write only store, but please enlighten me, what is the purpose of a write only store? Or more generically a store that holds a value that can never be read? Or am I misunderstanding?

It seems to me that the idea is to retrieve data only in a function-controlled way.
That you can retrieve data with, for example, $someStore.getConcreteData() and you can't read the entire $someData.
But I'll admit that I haven't used this. I write about it purely theoretically, because in Svelte there is such an option.

**Comment by zachlankton** [comment]
Ah gotcha, so for readonly, I'm thinking something like this:

```tsx
import { readonly } from '@rezact/rezact/signals';

let $writableStore = "someValue"
export const $readOnlyStore = readonly($writableStore);

$writableStore = "newValue"; // sets the store value to "newValue"
$readOnlyStore = "newValue"; // ERROR
```

For the other case, the best I can come up with is something like class private properties.  It sounds like this is more or less what you are after, some internal value that is managed through methods only, whether that is read or write.

Still not exactly sure the best way to approach this.  Perhaps you could give me an example of the code that you would like to be able to write to accomplish your vision?

**Reply by look997** [reply]
All I can say is that in Svelte 5 you just don't need to add a setter, and then you have this effect.
Because in Svelte 5 you have to set a getter/setter to make the value readable or assignable in a direct way.

**Reply by zachlankton** [reply]
I'm actively avoiding getters and setters. 

To be clear they are used under the hood, but the compiler transforms the code above into getters/setters so the developer does not have to.
